dask/array/gufunc.py

```diff
@@ -0,0 +1,343 @@
+
+    ## Split output
+    leaf_arrs = []
+    for i, cod, odt in zip(count(0), core_output_dimss, output_dtypes):
+        core_output_shape = tuple(core_shapes[d] for d in cod)
+        core_chunkinds = len(cod) * (0,)
+        output_shape = loop_output_shape + core_output_shape
+        output_chunks = loop_output_chunks + core_output_shape
+        leaf_name = "%s_%d-%s" % (name, i, token)
+        leaf_dsk = {(leaf_name,) + key[1:] + core_chunkinds: ((getitem, key, i) if nout else key) for key in keys}
+        leaf_arr = Array(sharedict.merge((leaf_name, leaf_dsk), dsk),
+                         leaf_name,
+                         chunks=output_chunks,
+                         shape=output_shape,
+                         dtype=odt)
+        leaf_arrs.append(leaf_arr)
+
+    return leaf_arrs if nout else leaf_arrs[0]
+
+
+@curry
+def gufunc(signature, func, **kwargs):
+    """
+    Apply a generalized ufunc [2]_ to arrays. The ``signature``
+    determines if the function consumes or produces core dimensions.
+    This method returns a wrapped function, which then can be bound
+    to arguments.
+
+    In other terms, this function is like np.vectorize, but for
+    the blocks of dask arrays. If the function itself shall also
+    be vectorized, ``vectorize=True`` can be used for convenience.
+
+    Parameters
+    ----------
+    signature: String
+        Specifies what core dimensions are consumed and produced by ``func``.
+        According to the specification of numpy.gufunc signature [2]_
+    func : callable
+        Function to call like ``func(*args, **kwargs)`` on input arrays
+        (``*args``) that returns an array or tuple of arrays. If multiple
+        arguments with non-matching dimensions are supplied, this function is
+        expected to vectorize (broadcast) over axes of positional arguments in
+        the style of NumPy universal functions [1]_ (if this is not the case,
+        set ``vectorize=True``).
+    *args : numpy/dask arrays or scalars
+        Arrays to which to apply the function. Core dimensions as specified in
+        ``signature`` need to come last.
+    output_dtypes : dtype or list of dtypes
+        dtype or list of output dtypes.
+    output_sizes : dict, optional
+        Optional mapping from dimension names to sizes for outputs. Only used if
+        new core dimensions (not found on inputs) appear on outputs.
+    vectorize: bool
+        If set to ``True``, ``np.vectorize`` is applied to ``func`` for
+        convenience. Defaults to ``False``.
+
+    Returns
+    -------
+    Wrapped function
+
+    Examples
+    --------
+    >>> import dask.array as da
+    ... import numpy as np
+    ... @da.gufunc("(i)->(),()", output_dtypes=2*(a.dtype,))
+    ... def stats(x):
+    ...     return np.mean(x, axis=-1), np.std(x, axis=-1)
+    ... a = da.random.normal(size=(10,20,30), chunks=5)
+    ... mean, std = stats(a)
+    ... mean.compute()  # doctest: +SKIP
```

Comment:
Ideally we would define this thing as a numpy gufunc and implement some protocol in dask to handle those when applied to dask arrays.  I suspect that this isn't doable today.  I'll write more about this in a longer comment.